File: Leetcode_Ds_from_scratch/array_and_string/groupAnagrams.java
package Leetcode_Ds_from_scratch.array_and_string;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class groupAnagrams {

    private static String getFrequencyString(String str){
        int[]freq = new int[26];
        for (char c:str.toCharArray()){
            freq[c-'a']++;
        }
        StringBuilder sb = new StringBuilder("");
       for(int i=0;i<26;i++){
        if(freq[i]>0){
            sb.append((char)('a'+i));
            sb.append(freq[i]);
        }
       }
        return sb.toString();
    }

    private static ArrayList ga(String[]arr){
        HashMap<String,List<String>> mymap = new HashMap<>();
        for(String s :arr){
          if(mymap.containsKey(getFrequencyString(s))){
            mymap.get(getFrequencyString(s)).add(s);
          }
          else{
            List<String> strlist  = new ArrayList<>();
            strlist.add(s);
            mymap.put(getFrequencyString(s), strlist);
          }
        }
        return new ArrayList<>(mymap.values());
    }
    public static void main(String[] args) {
        String[]strs = {"eat","tea","tan","ate","nat","bat"};
        System.err.println(ga(strs));
    }
}

File: 1_2026_PythonDsa/ProblemsOnArrays/GroupAnagrams.py
from collections import Counter
strs = ["eat","tea","tan","ate","nat","bat"]
'''
res=[]
for i in range(0,len(strs)):

    map1=Counter(strs[i])
    print(map1)
    curls=[]
    for j in range(i,len(strs)):
        if(map1==Counter(strs[j])):
            if(strs[j] not in curls):
                curls.append(strs[j])
    if(curls not in res):
        res.append(curls)
print(res)

'''
def getFrequency(s):
    freq=[0]*26
    for c in s:
        freq[ord(c)-ord("a")]+=1
    res=""
    for i in range(26):
        if freq[i] > 0:
            res += chr(ord('a') + i)
            res += str(freq[i])
    return res
mymap={}
for s in strs:
    key=getFrequency(s)
    if key in mymap:
        mymap[key].append(s)
    else:
        mymap[key]=[s]
print(mymap)



#mythought
mymap={}
for s in strs:
  
    x=Counter(s)
    key = tuple(sorted(x.items()))
    
    if key in mymap:
        mymap[key].append(s)
    else:
        mymap[key]=[s]
print(mymap.values())
